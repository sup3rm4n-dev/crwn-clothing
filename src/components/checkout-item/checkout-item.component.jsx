import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  clearItemFromCart,
  addItem,
  removeItem,
} from '../../redux/cart/cart.actions';

import {
  ImageContainer,
  TextContainer,
  QuantityContainer,
  RemoveButtonContainer,
  TableRow,
  TableCell,
  TableHeadCol,
} from './checkout-item.styles';

export const CheckOutItem = ({ cartItem, clearItem, addItem, removeItem }) => {
  const { name, imageUrl, price, quantity } = cartItem;
  const remove = () => removeItem(cartItem);
  const add = () => addItem(cartItem);
  const clear = () => clearItem(cartItem);

  return (
    <TableRow>
      <TableHeadCol scope='col' colspan='1'>
        <ImageContainer>
          <img src={imageUrl} alt={name} />
        </ImageContainer>
      </TableHeadCol>
      <TableCell>
        <TextContainer>{name}</TextContainer>
      </TableCell>
      <TableCell>
        <QuantityContainer>
          <div className='remove-item' onClick={remove}>
            &#10094;
          </div>
          <span>{quantity}</span>
          <div className='add-item' onClick={add}>
            &#10095;
          </div>
        </QuantityContainer>
      </TableCell>
      <TableCell>{price}</TableCell>
      <TableCell>
        <RemoveButtonContainer className='clear-item-button' onClick={clear}>
          &#10005;
        </RemoveButtonContainer>
      </TableCell>
    </TableRow>
  );
};

CheckOutItem.propTypes = {
  addItem: PropTypes.func.isRequired,
  cartItem: PropTypes.exact({
    id: PropTypes.number.isRequired,
    imageUrl: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    quantity: PropTypes.number.isRequired,
  }),
  clearItem: PropTypes.func.isRequired,
  removeItem: PropTypes.func.isRequired,
};

CheckOutItem.whyDidYouRender = false;

// const mapDispatchToProps = dispatch => ({
//   clearItem: item => dispatch(clearItemFromCart(item)),
//   addItem: item => dispatch(addItem(item)),
//   removeItem: item => dispatch(removeItem(item)),
// });

const mapDispatchToProps = {
  clearItem: item => clearItemFromCart(item),
  addItem: item => addItem(item),
  removeItem: item => removeItem(item),
};

export default connect(null, mapDispatchToProps)(CheckOutItem);
