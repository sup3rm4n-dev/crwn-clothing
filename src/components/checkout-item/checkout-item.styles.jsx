import styled from 'styled-components/macro';

import { media } from '../../helpers';

export const CheckoutItemContainer = styled.div`
  width: 100%;
  display: flex;
  min-height: 10rem;
  border-bottom: 1px solid darkgrey;
  padding: 1.5rem 0;
  font-size: 2rem;
  align-items: center;
  ${media.tablet`
    font-size: 1.8rem;
  `}
`;

export const TableRow = styled.tr`
  width: 100%;
  padding: 0;
  margin: 1rem 0;
  text-align: center;
  min-height: 10rem;
  border-bottom: 1px solid darkgrey;
  font-size: 2rem;
  align-items: center;
  ${media.tablet`
    font-size: 1.8rem;
  `}
`;

export const TableCell = styled.td`
  width: auto;
  margin: 0;
  padding: 0;
  min-height: 10rem;
  height: auto;
  text-align: center;
  &:first-child {
    text-align: left;
    width: 25%;
    padding-left: 0.5rem;
  }
  &:last-child {
    text-align: right;
    padding-right: 1.5rem;
  }
`;

export const TableHeadCol = styled.th`
  width: auto;
  min-height: 10rem;
  height: auto;
`;

export const ImageContainer = styled.div`
  width: auto;
  margin: 0;
  min-height: 100%;
`;

export const StyledImage = styled.img`
  width: 100%;
  height: 100%;
  vertical-align: middle;
  padding: 1rem 0;
  ${media.desktop`padding: 0.5rem 0;`}
  ${media.tablet`padding: 0.25rem 0;`}
  ${media.phone`padding: 0;`}
`;

export const TextContainer = styled.span`
  display: inline-block;
  text-align: center;
`;

export const QuantityContainer = styled.span`
  display: flex;
  align-items: center;
  justify-content: center;
  div {
    cursor: pointer;
    user-select: none;
  }
`;

export const QuantityButton = styled.div`
  cursor: pointer;
  user-select: none;
`;

export const Quantity = styled.span`
  margin: 0 1rem;
`;

export const RemoveButtonContainer = styled.div`
  cursor: pointer;
  text-align: right;
  display: inline-block;
  padding-right: 2.5rem;
  user-select: none;
  ${media.tablet`padding-right: 1.5rem;`}
  ${media.phablet`padding-right: 0;`}
`;
