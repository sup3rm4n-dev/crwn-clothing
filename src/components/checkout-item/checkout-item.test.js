import React from 'react';
import { shallow } from 'enzyme';
import { CheckOutItem } from './checkout-item.component';

describe('CheckOutItem component', () => {
  let wrapper;
  let mockClearItem;
  let mockAddItem;
  let mockRemoveItem;

  beforeEach(() => {
    mockClearItem = jest.fn();
    mockAddItem = jest.fn();
    mockRemoveItem = jest.fn();
    const mockProps = {
      cartItem: {
        imageUrl: 'www.testImage.com',
        price: 10,
        name: 'hats',
        quantity: 2,
      },
      clearItem: mockClearItem,
      addItem: mockAddItem,
      removeItem: mockRemoveItem,
    };

    wrapper = shallow(<CheckOutItem {...mockProps} />);
  });

  it('should render CheckOutItem component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call clearItem when remove button is clicked', () => {
    wrapper.find('.clear-item-button').simulate('click');
    expect(mockClearItem).toHaveBeenCalled();
  });

  it('should call addItem when right arrow is clicked', () => {
    wrapper.find('div.add-item').simulate('click');
    expect(mockAddItem).toHaveBeenCalled();
  });

  it('should call removeItem when left arrow is clicked', () => {
    wrapper.find('div.remove-item').simulate('click');
    expect(mockRemoveItem).toHaveBeenCalled();
  });
});
