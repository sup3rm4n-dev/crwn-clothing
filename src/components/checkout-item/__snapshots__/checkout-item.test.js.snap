// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckOutItem component should render CheckOutItem component 1`] = `ShallowWrapper {}`;
